File: src/main/scala/leetcode/com/LinkedListCycleII.scala
package leetcode.com

import scala.util.control.Breaks

class ListNode(var _x: Int = 0) {
  var next: ListNode = null
  var x: Int = _x
}


/**
 * 使用双指针方法
 */

object LinkedListCycleII {
  def detectCycle(head: ListNode): ListNode = {
    if (head == null || head.next == null) {
      return null
    }
    var slow: ListNode = head
    var fast: ListNode = head.next
    var hasCycle: Boolean = false

    Breaks.breakable {
      while (slow != null && fast != null) {
        if (slow.eq(fast)) {
          hasCycle = true
          Breaks.break
        }

        slow = slow.next

        if (fast.next == null) {
          return null
        }
        fast = fast.next.next
      }
    }

    if (hasCycle) {
      slow = head
      while (!slow.eq(fast)) {
        slow = slow.next
        fast = fast.next
      }
      return slow
    }
    null
  }
}

File: GoProject/leetcode/two—pointer/linked-list-cycle-ii.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {

	if head == nil || head.Next == nil || head.Next.Next == nil {
		return nil
	}

	fast, slow := head.Next.Next, head.Next

	for fast != slow {
		//  如果没有环
		if fast == nil || fast.Next == nil {
			return nil
		}
		fast = fast.Next.Next
		slow = slow.Next
	}

	fast = head

	for fast != slow {
		fast = fast.Next
		slow = slow.Next
	}
	return fast

}
